system/security/plugins/jwtSecurity/jwtSecurity.cpp

```diff
@@ -932,55 +945,70 @@ class CJwtSecurityManager : implements IDaliLdapConnection, public CBaseSecurity
+                // the just-cached user data from the earlier fetch/unpack/validate.
+                CriticalBlock   block(daliFetchCrit);
 
-                    decodeTokenFromDali(foundEncodedToken.str(), foundToken, foundRefreshToken);
+                userInfo = gUserCache.get(userName.str(), time(nullptr) - DALI_USER_CACHE_EXP_TIME_SECONDS);
 
-                    // Validate the token, which also caches it if
-                    // validation passes
-                    if (verifyToken(foundToken, userName.str(), foundRefreshToken, "", ""))
+                if (userInfo.get() == nullptr || !userInfo->isValid())
+                {
+                    // This secmgr plugin running within a Dali client connection
+                    // does not have accesss to user passwords so it cannot
+                    // authenticate against the JWT endpoint; if the token
+                    // is not in the cache or it is invalid then we have to
+                    // try to load it from Dali's key/value store
+                    StringBuffer        foundEncodedToken;
+
+                    if (daliStore.fetch(DALI_KVSTORE_STORE_NAME, DALI_KVSTORE_NAMESPACE, DALI_KVSTORE_KEY, foundEncodedToken, user, false))
                     {
-                        // Load parsed token out of cache
-                        userInfo = gUserCache.get(userName.str());
+                        std::string         foundToken;
+                        std::string         foundRefreshToken;
+
+                        decodeTokenFromDali(foundEncodedToken.str(), foundToken, foundRefreshToken);
+
+                        // Validate the token, which also caches it if
+                        // validation passes
+                        if (verifyToken(foundToken, userName.str(), foundRefreshToken, "", ""))
+                        {
+                            // Load parsed token out of cache
+                            userInfo = gUserCache.get(userName.str());
+                        }
+                        else
+                        {
+                            return SecAccess_Unavailable;
+                        }
                     }
                     else
                     {
+                        // No token in the K/V store;
                         return SecAccess_Unavailable;
                     }
                 }
-                else
-                {
-                    // No token in the K/V store;
-                    return SecAccess_Unavailable;
-                }
             }
 
-            if (strisame(key, "Scope"))
-                resultFlag = _authorizeEx(RT_FILE_SCOPE, *user, obj, nullptr, *userInfo);
-            else if (strisame(key, "workunit"))
-                resultFlag = _authorizeEx(RT_WORKUNIT_SCOPE, *user, obj, nullptr, *userInfo);
+            if (userInfo.get())
+            {
+                if (strisame(key, "Scope"))
```

Comment:
I'll have to remember to use strisame more often when I can.

Reply:
LOL Jake reminds me of that suite of functions all the time, as I keep forgetting they exist.